distributed/scheduler.py

```diff
@@ -4495,104 +4553,147 @@ def update_graph(
-                logger.exception(e)
+    def _pop_known_tasks(self, dsk: dict, dependencies: dict) -> set[str]:
+        # Avoid computation that is already finished
+        already_in_memory = set()  # tasks that are already done
+        for k, v in dependencies.items():
+            if v and k in self.tasks:
+                ts = self.tasks[k]
+                if ts.state in ("memory", "erred"):
+                    already_in_memory.add(k)
 
-        self.transitions(recommendations, stimulus_id)
+        done = set(already_in_memory)
+        if already_in_memory:
+            dependents = dask.core.reverse_dict(dependencies)
+            stack = list(already_in_memory)
+            while stack:  # remove unnecessary dependencies
+                key = stack.pop()
+                try:
+                    deps = dependencies[key]
+                except KeyError:
+                    deps = self.tasks[key].dependencies
+                for dep in deps:
+                    if dep in dependents:
+                        child_deps = dependents[dep]
+                    elif dep in self.tasks:
+                        child_deps = self.tasks[dep].dependencies
+                    else:
+                        child_deps = set()
+                    if all(d in done for d in child_deps):
+                        if dep in self.tasks and dep not in done:
+                            done.add(dep)
+                            stack.append(dep)
+        for anc in done:
+            dsk.pop(anc, None)
+            dependencies.pop(anc, None)
+        return done
 
-        for ts in touched_tasks:
-            if ts.state in ("memory", "erred"):
-                self.report_on_key(ts=ts, client=client)
+    @staticmethod
+    def materialize_graph(hlg) -> tuple[dict, dict, dict, dict]:
+        from distributed.worker import dumps_task
+
+        key_annotations = {}
+        dsk = dask.utils.ensure_dict(hlg)
+
+        for layer in hlg.layers.values():
+            if layer.annotations:
+                annot = layer.annotations
+                key_annotations.update({stringify(k): annot for k in layer})
+
+        dependencies, _ = get_deps(dsk)
+
+        # Remove `Future` objects from graph and note any future dependencies
+        dsk2 = {}
+        fut_deps = {}
+        for k, v in dsk.items():
+            dsk2[k], futs = unpack_remotedata(v, byte_keys=True)
+            if futs:
+                fut_deps[k] = futs
+        dsk = dsk2
+
+        # - Add in deps for any tasks that depend on futures
+        for k, futures in fut_deps.items():
+            dependencies[k].update(f.key for f in futures)
+        new_dsk = {}
+        # Annotation callables are evaluated on the non-stringified version of
+        # the keys
+        pre_stringified_keys = {}
+        for k, v in dsk.items():
+            new_k = stringify(k)
+            pre_stringified_keys[new_k] = k
+            new_dsk[new_k] = stringify(v, exclusive=hlg)
```

Comment:
Turns out this line is responsible for the most severe performance regressions since `hlg: HighLevelGraph` does not implement a fast path for `__contains__` such that `stringify` loops over the mapping for every check, i.e. this for-loop is secretly quadratic.

This can be easily fixed but from my POV this is another nail in the coffin for the tuples-as-tasks, c.f. https://github.com/dask/dask/issues/9969